### PR TITLE
E0060 and E0061 improvement

### LINKED ISSUE #35290: E0060 and E0061 bonus improvements for error messages
cc @GuillaumeGomez 

This issue tracks the bonus portions of issues #35215 and #35216. I worked on the `span_label` fixes for them and @jaredwy said he would take on the bonus portion.

**Bonus for E0060**:

>     error[E0060]: this function takes at least 1 parameter but 0 parameters were supplied
>       --> src/test/compile-fail/E0060.rs:16:14
>        |
>     16 |     unsafe { printf(); } //~ ERROR E0060
>        |              ^^^^^^^^ expected at least 1 parameter
>        |
>        = note: the following parameter type was expected: *const u8
> 
> Bonus: There may be a better way of displaying the information from the note. Could it be part of the label?

**Bonus for E0061**:

> 
> 
>     error[E0061]: this function takes 2 parameters but 1 parameter was supplied
>       --> src/test/compile-fail/E0061.rs:14:5
>        |
>     14 |     f(0); //~ ERROR E0061
>        |     ^^^^ expected 2 parameters
>        |
>        = note: the following parameter types were expected: u16, &str
> 
>     error: aborting due to previous error
> 
> Bonus: the types could be incorporated into the label rather than a note (though we may want to avoid types with long names in the label)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #35290.

r? @jonathandturner

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -2481,16 +2481,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-                                        if variadic {"at least "} else {""},
-                                        expected_count,
-                                        if expected_count == 1 {""} else {"s"}));
-
             let input_types = fn_inputs.iter().map(|i| format!("{:?}", i)).collect::<Vec<String>>();
             if input_types.len() > 0 {
-                err.note(&format!("the following parameter type{} expected: {}",
-                        if expected_count == 1 {" was"} else {"s were"},
-                        input_types.join(", ")));
+                err.span_label(sp, &format!("the following parameter type{} expected: {}",
```
Comment (jonathandturner) [inline src/librustc_typeck/check/mod.rs]: If we're going to make this a label, can we make it smaller?  Maybe something like it simply:

"parameter type{} expected: {}"

Makes we wish we had a span_label for the definition site, so they could see where the requirements were coming from (if we don't have that already)

Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: What about my proposition of list? A type a line if more than one type and it makes a nice output, easy to read. What do you think about it?

Comment (jonathandturner) [inline src/librustc_typeck/check/mod.rs]: A list in the label or in a note?  I think it would work better in the note.  Do you have an heuristic in mind?

Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: Not really. I can try both and show the result in here if you want.

Comment (jonathandturner) [inline src/librustc_typeck/check/mod.rs]: Showing here sounds good

**Comment by GuillaumeGomez** [comment]
First test with list in the error message directly:

```bash
./x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0061.rs
error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> src/test/compile-fail/E0061.rs:14:5
   |
14 |     f(0);
   |     ^^^^ the following parameter types were expected: 
* u16
* &str

error: aborting due to previous error
```

Not very beautiful.

**Comment by GuillaumeGomez** [comment]
And inside a note it gives:

```bash
./x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0061.rs
error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> src/test/compile-fail/E0061.rs:14:5
   |
14 |     f(0);
   |     ^^^^
   |
   = note: the following parameter types were expected: 
* u16
* &str

error: aborting due to previous error
```

Not very beautiful either...

I think that the errors should keep the indent if there is a newline. If you want I can add it @jonathandturner.

**Comment by jonathandturner** [comment]
Maybe:

```
./x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0061.rs
error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> src/test/compile-fail/E0061.rs:14:5
   |
14 |     f(0);
   |     ^^^^
   |
   = note: the following parameter types were expected:
   = note:   (u16, &str)
```

**Comment by jonathandturner** [comment]
Another alternative (after chatting on IRC) might be:

```
error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> src/test/compile-fail/E0061.rs:14:5
   |
11 | fn f(a: u16, b: &str) {}
   | - when calling this function
...
14 |     f(0);
   |     ^^^^ expected 2 parameters
   |
```

I wish we could point at the function name in the definition, but iirc, that span isn't yet available.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by jonathandturner** [comment]
Tidy failure:

```
/build/src/test/compile-fail/variadic-ffi-3.rs:20: tab character
```

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/test/compile-fail/issue-3044.rs [hunk 2]
```diff
@@ -15,6 +15,6 @@ fn main() {
     });
     //~^^ ERROR this function takes 2 parameters but 1 parameter was supplied
     //~| NOTE the following parameter types were expected
-    //~| NOTE expected 2 parameters
+    //~| NOTE _, _
```
Comment (jonathandturner) [inline src/test/compile-fail/issue-3044.rs]: Sigh, this one got worse.  I wish I'd fixed the type variable names also.  Once they're fixed, this should look better

**Comment by acrrd** [comment]
This broke the non bonus part of #35214

**Pull request merged**